### PR TITLE
PHP: Proto3 adoption

### LINKED ISSUE #6362: Deduplicate .proto files under src/php once PHP supports proto3.
Context: #4123

Once proto3 is supported, these files  should be removed and canonical versions from src/proto should be used.
```
src/php/tests/generated_code/math.proto
src/php/tests/interop/test.proto
src/php/tests/interop/empty.proto
src/php/tests/interop/messages.proto
```

### LINKED ISSUE #6637: PHP protoc plugin
Implement the gRPC PHP protoc plugin, after Proto3 PHP support has been rolled out.

### LINKED ISSUE #8201: grpc 1.0.0 does not build with protobuf 3.1.0
I think grpc 1.0.0 ships with some proto headers that should better be auto-generated:
```
[CXX]     Compiling src/cpp/ext/reflection.pb.cc
In file included from src/cpp/ext/reflection.grpc.pb.cc:39:0:
include/grpc++/ext/reflection.pb.h:51:2: error: #error This file was generated by an older version of protoc which is
 #error This file was generated by an older version of protoc which is
  ^
include/grpc++/ext/reflection.pb.h:52:2: error: #error incompatible with your Protocol Buffer headers. Please
 #error incompatible with your Protocol Buffer headers.  Please
  ^
include/grpc++/ext/reflection.pb.h:53:2: error: #error regenerate this file with a newer version of protoc.
 #error regenerate this file with a newer version of protoc.
```

### PR DESCRIPTION
**Pull request by stanley-cheung**
Fixes #6637 
Fixes #6362
Fixes #8201 
Step towards #5983

Key changes:
 * Added the gRPC PHP Plugin to generate client service stub classes
 * Upgraded protobuf dependency to v3.1.0 [a428e42](https://github.com/google/protobuf/releases)
 * Removed the previous `stanley-cheung/Protobuf-PHP` library and ruby dependency but if user decides to continue to use that library things will continue to work
 * Updated the Math generated code test, and interop tests
 * Removed the duplicated sources `math.proto`, `test.proto`, `empty.proto`, `messages.proto`. Now we can share the sources with `src/protos`
 * Updated jenkins PHP dockerfiles to use this new library, can now remove the ruby dependency.
 * Updated README
 * No change to the PECL extension

Notes:
 * To test code generation, please run `src/php/bin/generate_proto_php.sh`.
 * From user's perspective, this is almost a drop-in replacement of the previous `Protobuf-PHP` library. See the changes in `src/php/tests/generated_code/AbstractGeneratedCodeTest.php`. Aside from a different code generator, all we need to change is to include the re-generated stub code, and the client code continues to work.
 * Although `google/protobuf` don't officially support PHP7 yet because the PECL extension isn't compatible yet, the composer version will actually run on PHP 5.5, 5.6 and 7.0.

Next steps:
 * Update the `examples/php` directory helloworld greeter and route guide example. But that will need our `grpc/grpc` version 1.1 to be released to Packagist/composer first.
 * Update grpc.github.io website and change any reference to the old `Protobuf-PHP` library.

### CODE REVIEW

File: templates/tools/dockerfile/stress_test/grpc_interop_stress_php/Dockerfile.template [hunk 1]
```diff
@@ -37,25 +37,10 @@
...
-  # install of Protobuf extensions
-  #
-  # rake: a ruby version of make used to build the PHP Protobuf extension
-  RUN /bin/bash -l -c "rvm all do gem install ronn rake"
-  
   # Install composer
   RUN curl -sS https://getcomposer.org/installer | php
   RUN mv composer.phar /usr/local/bin/composer
   
-  # Download the patched PHP protobuf so that PHP gRPC clients can be generated
```
Comment (jtattermusch) [inline templates/tools/dockerfile/stress_test/grpc_interop_stress_php/Dockerfile.template]: Glad we are finally getting rid of this :-)

File: src/php/README.md [hunk 2]
```diff
@@ -98,44 +128,52 @@ $ ./bin/run_tests.sh
...
 
 ```sh
-$ git clone https://github.com/stanley-cheung/Protobuf-PHP
-$ cd Protobuf-PHP
-$ gem install rake ronn
-$ rake pear:package version=1.0
-$ sudo pear install Protobuf-1.0.tgz
+$ git clone https://github.com/grpc/grpc
+$ make plugins
+$ sudo make install-plugins
```
Comment (jtattermusch) [inline src/php/README.md]: add a comment in sense of "only do make install-plugins if necessary and you cannot use the plugins in bins/opt/.. There is no supported way of uninstalling".

File: src/php/bin/generate_proto_php.sh [hunk 3]
```diff
@@ -28,13 +28,32 @@
...
+sed -i 's/message Empty/message EmptyMessage/g' \
+    src/proto/grpc/testing/empty.proto
+sed -i 's/grpc\.testing\.Empty/grpc\.testing\.EmptyMessage/g' \
+    src/proto/grpc/testing/test.proto
 
-protoc-gen-php -i $gen_code -o $gen_code $gen_code/math.proto
+protoc --proto_path=. \
+       --php_out=src/php/tests/interop \
+       --grpc_out=src/php/tests/interop \
+       --plugin=protoc-gen-grpc=$(which grpc_php_plugin) \
```
Comment (jtattermusch) [inline src/php/bin/generate_proto_php.sh]: you are still using which here.

File: third_party/protobuf [hunk 4]
```diff
@@ -1 +1 @@
-Subproject commit bba446bbf2ac7b0b9923d4eb07d5acd0665a8cf0
+Subproject commit 8c88762e3ef78b81435ecfba973c2e9d3b74f13b
```
Comment (jtattermusch) [inline third_party/protobuf]: you'll also need to update this in the sanity test, otherwise sanity test will fail.

Comment (stanley-cheung) [inline third_party/protobuf]: Fixed

File: src/php/README.md (hunk 2 above)
Comment (stanley-cheung) [inline src/php/README.md]: I removed the `make plugins` and just mention that the plugin will be found in the `bins/opt` directory. But the ideal solution is still to provide the package for the plugin.

File: templates/tools/dockerfile/stress_test/grpc_interop_stress_php/Dockerfile.template (hunk 1 above)
Comment (stanley-cheung) [inline templates/tools/dockerfile/stress_test/grpc_interop_stress_php/Dockerfile.template]: :)

File: src/php/bin/generate_proto_php.sh (hunk 3 above)
Comment (stanley-cheung) [inline src/php/bin/generate_proto_php.sh]: Didn't realize there's a second occurrence. Fixed.

**Comment by stanley-cheung** [comment]
@murgatroid99 @nicolasnoble This is ready for a broader review. All relevant tests (sanity, PHP/PHP7 tests, interop, portability) passed. Can you guys also review this please? This is intended for a 1.1 release of gRPC and will introduce a dependency on protobuf @ v3.1.0. Thanks.

**Comment by jtattermusch** [comment]
I am happy with the changes. Leaving final LGTM to @murgatroid99  or @nicolasnoble.

**Pull request merged**